### PR TITLE
Docker nodejs16 upgrade

### PR DESCRIPTION
**Pull request by martadinata666**
1. Use `ubuntu:20.04 + nodejs 16` base
2. Use of 22.04 need further identification, see https://github.com/Tzahi12345/YoutubeDL-Material/issues/608
3. Clean up some packages
4. Restructure fetching backend and building frontend
5. Rename docker-build.sh to ffmpeg-fetch.sh to prevent confusion naming
6. Add --raw to make pm2 show logs as it
7. Using `ffmpeg-fetch.sh` for compability with `docker build` method

### CODE REVIEW

**Comment by GlassedSilver** [comment]
Switching to Debian breaks my heart after all this work to get Ubuntu done mere days after pulling two all-nighters to look into all sorts of stuff. (it was my first time working on a Dockerfile in such a profound way)

Any reason why it has to be Debian rather than Ubuntu?

**Comment by GlassedSilver** [comment]
https://github.com/Tzahi12345/YoutubeDL-Material/commit/b2d8c4ef55219512206aad49a097129917d52f4b deliberately removed pm2 logging, not sure why, but it's not coincidence it seems.

Some other changes might make sense to be documented decisions as well. I'm glad you added more reasoning than some might provide for their changes, but I'd like to know why we bring in the volume in the dockerfile? (did we have a broken dockerfile config before and didn't catch some known issues because of it? Just legitimately pondering, since you seem more knowledgable about Dockerfile creation than I am)

Any reason why you use apt instead of apt-get which is said to be better for background automation because its functionality is as close to promised to not change as possible?

Oh and now that I'm looking at the ffmpeg acquiring script again I'm not even sure why we deal with all that cleanup in multi-stage build. (both you and me that is :P)

I mean all of it is discarded anyways except for the things we specifically copy out of it, if I'm not mistaken, right? the building VM will wipe it all in the end anyhow. Just something I spotted now that I looked at the file again.

**Comment by martadinata666** [comment]
It just personal preferences, if we go ubuntu also will do. Not really must use debian.
```
FROM node:16-bullseye-slim`
```
to 
```
FROM ubuntu:22.04 
RUN apt update && \
        apt install -y --no-install-recommends && \
        curl -fsSL https://deb.nodesource.com/setup_16.x | bash - && \
        apt install -y --no-install-recommends nodejs ...
```

**Comment by Tzahi12345** [comment]
I put up #610 as a hotfix for #608. If this works without issue we can merge it as it looks cleaner than our current Dockerfile, but I just wanted to get the current one working for everyone.

**Review by GlassedSilver — COMMENTED** [review]
ok I guess? Haven't tested, but Dockerfile looks plausible :)

**Comment by martadinata666** [comment]
The image result was here `martadinata666/youtubedl-material:nodejs16-test` should be compatible with our docker-compose.yml, pretty confidence on amd64/arm64, i dont have armv7 devices, cant test it.

and some catch about mongodb, arm64 system like RPI cant use mongodb:latest as it missing cpu feature `avx something` to run it, as RPI use mongodb:4.4 , not really sure how to deal with this.

**Review by Tzahi12345 — DISMISSED** [review]

**Comment by martadinata666** [comment]
There is some combination that we must identify
1. 22.04 + node12 (borked)
2. 20.04 + node12 (work)
3. 22.04 + node12 (not preferable due nodejs EOL)
4. 22.04 + node16 (not publicly tested yet) (tested on my local debian 11 and ubuntu 20.04 machine works)

#610 also lead to #596 the TS said he use debian 11, i also tested on 22.04+node16 on debian 11, it works.

So i think, for now we stay at 20.04+node16, for best bet.

File: Dockerfile [hunk 1]
```diff
@@ -1,75 +1,62 @@
...
 
-COPY docker-build.sh .
-RUN sh ./docker-build.sh
-
-#--------------# Stage 2
-
-FROM ubuntu:22.04 as frontend
-
+# Create our Ubuntu 22.04 with node 16
+FROM ubuntu:22.04 AS base
```
[the remark this answers is not in the document]
Comment (GlassedSilver) [reply Dockerfile]: This got marked as resolved but the PR still has 22.04 if I'm not mistaken and there's no explanation what may have been considered that renders 22.04 good to go?

Sorry for the back and forth here, @martadinata666, your contributions are most certainly valued and appreciated. 👍

[the remark this answers is not in the document]
Comment (GlassedSilver) [reply Dockerfile]: Ah ok, should have looked in the comment below I guess

**Comment by GlassedSilver** [comment]
> There is some combination that we must identify
> 
>     1. 22.04 + node12 (borked)
> 
>     2. 20.04 + node12 (work)
> 
>     3. 22.04 + node12 (not preferable due nodejs EOL)
> 
>     4. 22.04 + node16 (not publicly tested yet) (tested on my local ubuntu 22.04 arm64 and ubuntu 20.04 amd64 machine works)
> 
> 
> #610 also lead to #596 the TS said he use debian 11, tested with dietpi based on debian 11 amd64 also works.
> 
> So i think, for now we stay at 20.04+node16, for best bet.

Sounds good to me, I'll give this a green light then and we'll see how it fares in the field. :)

File: Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (martadinata666) [reply Dockerfile]: only need change the 20.04 base, as the ffmpeg fetcher let it be 22.04 as it doest really matter. And that is the #610 PR link, mine was here https://github.com/martadinata666/YoutubeDL-Material/blob/a57a25133c0f5e893074127a4a2ef5e818481cb0/Dockerfile#L19-L24

**Comment by GlassedSilver** [comment]
Well now the Dockerfile still needs to reference the script if you keep the script instead of integrating the process into the Dockerfile already :P

Other than that, seems good to me? I guess @Tzahi12345 would need to look at it again and give his opinion and then we're ready to merge I think.

**Comment by GlassedSilver** [comment]
The Dockerfile is still missing changes though? o.O

**Comment by martadinata666** [comment]
> Well now the Dockerfile still needs to reference the script if you keep the script instead of integrating the process into the Dockerfile already :P
> 
> Other than that, seems good to me? I guess @Tzahi12345 would need to look at it again and give his opinion and then we're ready to merge I think.

The usage of script is usefull for personal docker build using `docker build` as `docker buildx` got `TARGETPLATFORM` to check.
Should we stay use `ffmpeg-fetch.sh` script?

**Comment by GlassedSilver** [comment]
> > Well now the Dockerfile still needs to reference the script if you keep the script instead of integrating the process into the Dockerfile already :P
> > Other than that, seems good to me? I guess @Tzahi12345 would need to look at it again and give his opinion and then we're ready to merge I think.
> 
> The usage of script is usefull for personal docker build using `docker build` as `docker buildx` got `TARGETPLATFORM` to check. Should we stay use `ffmpeg-fetch.sh` script?

I say we keep using the script, it's very relevant to the project and not docker-specific, so it makes sense to keep it in a general repo-context. 👍

**Review by GlassedSilver — APPROVED** [review]
This gets my go ahead.

Since the FFmpeg acquisition can basically run in whatever I'll feel free to green light this solo.
The requested change to switch to Ubuntu 20.04 is wouldn't affect the final docker build since the base does make use of 20.04 and that's what ships to the end user. 👍

**Pull request merged**